### PR TITLE
Add body3d dataset

### PR DESCRIPTION
**Pull request by ly015**
* add dataset interface
* add human3.6m dataset for 3d keypoint detection

### CODE REVIEW

File: mmpose/datasets/datasets/__init__.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from ...deprecated import (TopDownFreiHandDataset, TopDownOneHand10KDataset,
                            TopDownPanopticDataset)
+from .body3d import Body3DH36MDataset
```
Comment (innerlee) [inline mmpose/datasets/datasets/__init__.py:3]: Body or Pose? What do you think @jin-s13

Comment (jin-s13) [reply mmpose/datasets/datasets/__init__.py:3]: Body is better.

File: mmpose/datasets/datasets/body3d/body3d_base_dataset.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+from torch.utils.data import Dataset
+
+from mmpose.datasets.pipelines import Compose
+
+
+class Body3DBaseDataset(Dataset, metaclass=ABCMeta):
+    """Base class for 3D human pose datasets."""
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_base_dataset.py:11]: docstring for Args

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        10: 'head',
+        11: 'left_shoulder',
+        12: 'left_elbow',
+        13: 'left_wrist',
+        14: 'right_shoulder',
+        15: 'right_elbow',
+        16: 'right_wrist'
+
+    Args:
+        PoseTransferBaseDataset ([type]): [description]
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:41]: Where is this come from

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    """
+
+    JOINT_NAMES = [
+        'Root', 'LHip', 'LKnee', 'LFoot', 'RHip', 'RKnee', 'RFoot', 'Spine',
+        'Thorax', 'NeckBase', 'Head', 'LShoulder', 'LElbow', 'LWrist',
+        'RShoulder', 'RElbow', 'RWrist'
+    ]
+
+    JOINT_IDX_GLOBAL = [
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:50]: How to understand this var name

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        # get 2D joints
+        if self.joint_2d_src == 'gt':
+            assert data_info['joint_2d'].shape[1] == self.JOINT_NUM_GLOBAL
+            data_info['joint_2d'] = data_info['joint_2d'][:, self.
+                                                          JOINT_IDX_GLOBAL]
+        elif self.joint_2d_src == 'detection':
+            data_info['joint_2d'] = self._load_joint_2d_detection(
+                self.joint_2d_det_file)
+        elif self.joint_2d_src == 'pipeline':
+            pass
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:107]: Add an assertion on the presence of 'joint_2d'

File: mmpose/datasets/datasets/body3d/body3d_base_dataset.py (hunk 2 above)
Comment (ly015) [reply mmpose/datasets/datasets/body3d/body3d_base_dataset.py:11]: Will add a docstring.

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+                _indices[i:(i + _len):_step]
+                for i in range(0, n_frame - _len + 1)
+            ]
+            sample_indices.extend(seqs_from_video)
+
+        return sample_indices
+
+    def evaluate(self, outputs, res_folder, metric='joint_error', **kwargs):
+        metrics = metric if isinstance(metric, list) else [metric]
+
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:156]: Try to bound the kwargs, like

assert len(kwargs) == 0

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    def _load_camera_param(self, camear_param_file):
+        """Load camera parameters from file."""
+        return mmcv.load(camear_param_file)
+
+    def get_camera_param(self, imgname):
+        """Get camera parameters of a frame by its image name."""
+        assert hasattr(self, 'camera_param')
+        subj, _, camera = self._parse_h36m_imgname(imgname)
+        print(imgname)
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:252]: debugging code?

**Comment by innerlee** [comment]
Test coverage of the new dataset is low (< 80%)

image

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py (hunk 6 above)
Comment (ly015) [reply mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:156]: I thought that the purpose of the "kwargs" here is to allow arbitrary args passed in but not used. It seems that "kwargs" is not bounded in other dataset classes' evaluate method.

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py (hunk 5 above)
Comment (innerlee) [reply mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:118]: Oh I misunderstood the code. Simply ignore it :)

There is no need to add assertion in L110

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py (hunk 6 above)
Comment (innerlee) [reply mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:175]: Yeah this is the source of bugs. It is dangerous to omit arbitrary args. I am advocating to bound the kwargs, and it is still in-progress.

Comment (ly015) [reply mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:175]: Then maybe just remove **kwargs from the argument list?

File: mmpose/datasets/datasets/body3d/body3d_base_dataset.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+import numpy as np
+from torch.utils.data import Dataset
+
+from mmpose.datasets.pipelines import Compose
+
+
+class Body3DBaseDataset(Dataset, metaclass=ABCMeta):
+    """Base class for 3D human pose datasets.
+
+    Subclasses should consider everwriting following methods:
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_base_dataset.py:13]: everwriting

File: mmpose/datasets/datasets/body3d/body3d_base_dataset.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        - build_sample_indices
+        - evaluate
+
+    Args:
+        ann_file (str): Path to the annotation file.
+        img_prefix (str): Path to a directory where images are held.
+            Default: None.
+        data_cfg (dict): config
+        pipeline (list[dict | callable]): A sequence of data transforms.
+        test_mode (bool): Store True when building test or
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_base_dataset.py:25]: del `Store`

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        'Root', 'LHip', 'LKnee', 'LFoot', 'RHip', 'RKnee', 'RFoot', 'Spine',
+        'Thorax', 'NeckBase', 'Head', 'LShoulder', 'LElbow', 'LWrist',
+        'RShoulder', 'RElbow', 'RWrist'
+    ]
+
+    # The joint indices in the annotation file
+    JOINT_IDX_ANNOTATION = [
+        14, 2, 1, 0, 3, 4, 5, 16, 12, 17, 18, 9, 10, 11, 8, 7, 6
+    ]
+    # The total joint number in the annotatino file
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:61]: annotatino

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    JOINT_NUM_ANNOTATION = 24
+
+    # 2D joint source options:
+    # "gt": from the annotation file
+    # "detection": from a detection result file of 2D keypoint
+    # "pipeline": will be generate in the pipeline
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:67]: `will be generate in the pipeline` -> `generated by the pipeline`

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+            joint_error.append(error)
+            joint_error_pa.append(error_pa)
+
+        name_value_tuples = [('MPJPE', np.mean(joint_error)),
+                             ('MPJPE-PA', np.mean(joint_error_pa))]
+
+        return name_value_tuples
+
+    @staticmethod
+    def _joint_error_kernel(pred_joints_3d, gt_joints_3d, joints_3d_visible):
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:235]: rename to `MPJPE`  ?

Comment (jin-s13) [reply mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:235]: Better move to `mmpose/core/evaluation`

**Review by zengwang430521 — APPROVED** [review]

File: mmpose/datasets/datasets/body3d/body3d_base_dataset.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        ann_info['upper_body_ids'] = None
+        ann_info['lower_body_ids'] = None
+        ann_info['joint_weights'] = np.full(
+            self.num_joints, 1.0, dtype=np.float32)
+
+        self.ann_info.update(ann_info)
+
+    def load_annotations(self):
+        """Load data annotation."""
+        data = np.load(self.ann_file)
```
Comment (innerlee) [inline mmpose/datasets/datasets/body3d/body3d_base_dataset.py:79]: Are all annotation files numpy formats

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+                kpts.append({
+                    'keypoints': preds[i],
+                    'target_id': target_id,
+                })
+
+        mmcv.dump(kpts, res_file)
+
+        name_value_tuples = []
+        for _metric in metrics:
+            if _metric == 'joint_error':
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:200]: how about `_metric == 'mpjpe':`

File: mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+    JOINT_NUM_ANNOTATION = 24
+
+    # 2D joint source options:
+    # "gt": from the annotation file
+    # "detection": from a detection result file of 2D keypoint
+    # "pipeline": will be generate by the pipeline
+    SUPPORTED_JOINT_2D_SRC = {'gt', 'detection', 'pipeline'}
+
+    # metric
+    ALLOWED_METRICS = {'joint_error'}
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_h36m_dataset.py:71]: joint_error -> mpjpe

**Pull request merged**